### PR TITLE
Rectify: Grep BRE Syntax Immunity

### LINKED ISSUE #1184: Grep tool uses ripgrep but models emit Bash grep OR-syntax, causing silent search failures
## Problem

The `Grep` tool in Claude Code is backed by ripgrep, but models (particularly Sonnet) frequently emit Bash grep OR-syntax (`\|`) instead of ripgrep syntax (`|` or `(a|b)`). Ripgrep interprets `\|` as a literal backslash followed by a pipe, causing searches to silently return zero results when matches exist.

This was observed in session `a268d30c-3938-45d5-b145-07a583b357c9` (make-plan for #1164), where it amplified two separate exploration rabbit holes by 15+ turns each. The model searched for functions that exist but got zero results, concluded the functions didn't exist, and kept searching with different (also broken) patterns.

### Specific examples from the session

**Example 1** — Searching for contract generation functions in `contracts.py`:
```
Grep pattern: "def.*generate\|def.*regenerate\|def.*build.*contract\|write.*contracts"
Expected:     matches `generate_recipe_card` in contracts.py
Actual:       0 results (ripgrep treats \| as literal)
```

**Example 2** — Searching for capture-related functions:
```
Grep pattern: "def.*capture|resolve_capture|extract_capture"  
```
(Mixed — some `|` correct, some work by accident because they're at word boundaries)

**Example 3** — Searching for staleness functions:
```
Grep pattern: "contract.*generate|generate.*contract|regenerate|write.*contract|staleness_cache.*build"
```
(This one used correct `|` syntax and worked — showing the model is inconsistent)

### Impact

When a Grep call silently returns empty, the model interprets it as "this code doesn't exist" rather than "my search syntax was wrong." This triggers exploration loops where the model:
1. Searches with broken syntax → 0 results
2. Concludes the function/pattern doesn't exist
3. Searches for alternative explanations
4. Searches with a different (also broken) pattern → 0 results
5. Repeats, consuming 10-20 turns per rabbit hole

In session `a268d30c`, two rabbit holes were directly amplified by this:
- **Contract regeneration search** (turns 62-78): `generate_recipe_card` exists in `contracts.py` but was never found due to `\|` syntax. The model spent 15+ turns searching for how contracts are regenerated.
- **Capture semantics search** (turns 22-41): Multiple searches for capture-related functions used mixed syntax, with some failing silently.

### Root Cause

This is a model behavior pattern — Sonnet (and likely other models) learned Bash grep syntax from training data where `grep "pattern1\|pattern2"` is the standard OR idiom. The Grep tool wraps ripgrep, which uses a different regex dialect where `|` is the alternation operator without escaping. There is no validation or warning when a pattern contains `\|`.

## Proposed Fix

Add a lightweight pattern lint to the Grep tool's PreToolUse hook that detects `\|` in the pattern argument and either:

**Option A — Auto-correct:** Replace `\|` with `|` before passing to ripgrep, with a debug log noting the correction.

**Option B — Warn and proceed:** Emit a warning in the tool result like "Note: `\|` in ripgrep is a literal backslash-pipe, not alternation. Did you mean `|`?" so the model self-corrects on the next call.

**Option C — Skill-level instruction:** Add a note to CLAUDE.md section 3.3 or skill preambles: "The Grep tool uses ripgrep syntax. Use `|` for alternation, not `\|` (Bash grep syntax)."

Options A and B are defensive and work regardless of which skill or model is running. Option C relies on model attention to instructions.

## Acceptance Criteria

- [ ] `\|` in Grep patterns is either auto-corrected or flagged with a warning
- [ ] Test: Grep pattern containing `\|` either produces correct alternation results or returns a visible warning
- [ ] No false positives for patterns where `\|` is intentionally a literal match (edge case — very rare in code search)

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

The `Grep` tool wraps ripgrep (ERE: `|` for alternation), but models trained on Bash data emit `\|` (POSIX BRE), causing silent 0-result failures. The architecture has zero runtime interception for this failure mode. Part A installs the runtime immunity layer: a `PreToolUse` hook that denies `Grep` calls with `\|` in their pattern and returns a corrected pattern suggestion. This makes the failure **loud, instructive, and self-correcting** at the exact tool boundary where the bug manifests, eliminating the silent-failure rabbit-hole entirely.

Part B (separate task) will address the upstream source: SKILL.md content lint (pre-flight CI enforcement) and replacing BRE `grep` patterns in skill files with ripgrep-idiomatic `rg` syntax.

### Error/Resilience Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 40, 'rankSpacing': 55, 'curve': 'basis'}}}%%
flowchart TB
    %% CLASS DEFINITIONS %%
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef gap fill:#ff6f00,stroke:#ffa726,stroke-width:2px,color:#000;
    classDef integration fill:#c62828,stroke:#ef9a9a,stroke-width:2px,color:#fff;

    %% TERMINAL NODES %%
    T_ALLOW([ALLOWED<br/>pass-through])
    T_DENY([DENIED<br/>correction surfaced])
    T_ERR_FINDING([ERROR Finding<br/>recipe validation])
    T_TEST_FAIL([TEST FAILURE<br/>static corpus check])
    T_SILENT_FAIL([SILENT FAILURE<br/>0 results, no error])

    %% ─────────────────────────────────────────────────────── %%
    subgraph Problem ["PROBLEM: SILENT FAILURE (root cause)"]
        RIPGREP["ripgrep (ERE)<br/>━━━━━━━━━━<br/>| = alternation<br/>\\| = literal backslash-pipe"]
        BRE_PATTERN["Model emits \\| in Grep pattern<br/>━━━━━━━━━━<br/>Trained on Bash/POSIX BRE data<br/>grep uses \\| for alternation"]
        SILENT["0 results returned<br/>━━━━━━━━━━<br/>No exception, no warning<br/>Correct code NOT found"]
    end

    %% ─────────────────────────────────────────────────────── %%
    subgraph Layer1 ["LAYER 1: Runtime PreToolUse Hook (★ NEW)"]
        HOOK_ENTRY["★ grep_pattern_lint_guard.py<br/>━━━━━━━━━━<br/>PreToolUse on Grep tool<br/>Registered in ● hook_registry.py"]

        subgraph FailOpen ["Fail-Open Gates"]
            G_JSON{"JSON<br/>parse<br/>ok?"}
            G_TOOL{"tool_name<br/>== Grep?"}
            G_PATTERN{"pattern is<br/>str with \\|?"}
        end

        DENY_ACTION["Deny + Correct<br/>━━━━━━━━━━<br/>replace \\| → |<br/>permissionDecision: deny<br/>corrected pattern in reason"]
    end

    %% ─────────────────────────────────────────────────────── %%
    subgraph Layer2 ["LAYER 2: Recipe Validation Rule (● MODIFIED)"]
        RULE_ENTRY["● rules_skill_content.py<br/>━━━━━━━━━━<br/>★ grep-bre-alternation-in-skill<br/>Severity: ERROR"]

        subgraph SkipGates ["Skip Gates (Fail-Open)"]
            G_TOOL2{"tool ==<br/>run_skill?"}
            G_SKILL{"skill_name<br/>resolved?"}
            G_SKILL_MD{"SKILL.md<br/>exists?"}
            G_BASH{"bash blocks<br/>present?"}
        end

        G_GIT_GREP{"--grep=<br/>context?<br/>(git BRE)"}
        GREP_RE["★ _GREP_BRE_ALTERNATION_RE<br/>━━━━━━━━━━<br/>regex: grep...quoted\\|pattern<br/>not preceded by = or -"]
        RULE_FINDING["RuleFinding: ERROR<br/>━━━━━━━━━━<br/>skill name, line count<br/>fix hint: rg 'foo|bar'"]
    end

    %% ─────────────────────────────────────────────────────── %%
    subgraph Layer3 ["LAYER 3: Static Corpus Contract (★ NEW)"]
        TEST_ENTRY["★ test_skill_tool_syntax_contracts.py<br/>━━━━━━━━━━<br/>Parametrized over all SKILL dirs<br/>Scans bash blocks at test time"]
        G_GIT_GREP2{"--grep=<br/>context?<br/>(git BRE)"}
        TEST_VIOLATIONS["Collect violations<br/>━━━━━━━━━━<br/>Line-level grep BRE matches<br/>Excluded: git --grep= BRE"]
    end

    %% ─────────────────────────────────────────────────────── %%
    subgraph Registry ["HOOK REGISTRY (● MODIFIED)"]
        HOOK_DEF["● HOOK_REGISTRY entry<br/>━━━━━━━━━━<br/>matcher: Grep<br/>scripts: grep_pattern_lint_guard.py"]
        HASH_CHECK["registry.sha256<br/>━━━━━━━━━━<br/>● Updated hash digest<br/>Drift detection gate"]
    end

    %% ─────────────────────────────────────────────────────── %%
    %% PROBLEM FLOW %%
    BRE_PATTERN -->|"without guard"| RIPGREP
    RIPGREP --> SILENT
    SILENT --> T_SILENT_FAIL

    %% LAYER 1 FLOW %%
    BRE_PATTERN -->|"with guard active"| HOOK_ENTRY
    HOOK_ENTRY --> G_JSON
    G_JSON -->|"JSONDecodeError / OSError"| T_ALLOW
    G_JSON -->|"ok"| G_TOOL
    G_TOOL -->|"not Grep"| T_ALLOW
    G_TOOL -->|"Grep"| G_PATTERN
    G_PATTERN -->|"no \\| present"| T_ALLOW
    G_PATTERN -->|"\\| detected"| DENY_ACTION
    DENY_ACTION --> T_DENY

    %% REGISTRY %%
    HOOK_DEF --> HOOK_ENTRY
    HOOK_DEF --> HASH_CHECK

    %% LAYER 2 FLOW %%
    RULE_ENTRY --> G_TOOL2
    G_TOOL2 -->|"other tool"| T_ALLOW
    G_TOOL2 -->|"run_skill"| G_SKILL
    G_SKILL -->|"not resolved"| T_ALLOW
    G_SKILL -->|"resolved"| G_SKILL_MD
    G_SKILL_MD -->|"missing / OSError"| T_ALLOW
    G_SKILL_MD -->|"found"| G_BASH
    G_BASH -->|"no bash blocks"| T_ALLOW
    G_BASH -->|"has blocks"| GREP_RE
    GREP_RE --> G_GIT_GREP
    G_GIT_GREP -->|"--grep= context"| T_ALLOW
    G_GIT_GREP -->|"not git context"| RULE_FINDING
    RULE_FINDING --> T_ERR_FINDING

    %% LAYER 3 FLOW %%
    TEST_ENTRY --> G_GIT_GREP2
    G_GIT_GREP2 -->|"--grep= context"| T_ALLOW
    G_GIT_GREP2 -->|"grep \\| in bash"| TEST_VIOLATIONS
    TEST_VIOLATIONS --> T_TEST_FAIL

    %% CLASS ASSIGNMENTS %%
    class RIPGREP,BRE_PATTERN integration;
    class SILENT gap;
    class HOOK_ENTRY,DENY_ACTION newComponent;
    class G_JSON,G_TOOL,G_PATTERN detector;
    class RULE_ENTRY,GREP_RE newComponent;
    class G_TOOL2,G_SKILL,G_SKILL_MD,G_BASH,G_GIT_GREP detector;
    class RULE_FINDING gap;
    class TEST_ENTRY newComponent;
    class G_GIT_GREP2 detector;
    class TEST_VIOLATIONS gap;
    class HOOK_DEF,HASH_CHECK handler;
    class T_ALLOW,T_DENY,T_ERR_FINDING,T_TEST_FAIL,T_SILENT_FAIL terminal;
```

### Module Dependency Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 50, 'rankSpacing': 70, 'curve': 'basis'}}}%%
graph TB
    %% CLASS DEFINITIONS %%
    classDef cli fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef gap fill:#ff6f00,stroke:#ffa726,stroke-width:2px,color:#000;
    classDef integration fill:#c62828,stroke:#ef9a9a,stroke-width:2px,color:#fff;
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;

    %% ─── LAYER 0 / PACKAGE ROOT ─────────────────────────────────────────── %%
    subgraph L0 ["L0 — FOUNDATION / PACKAGE ROOT"]
        direction LR
        CORE["core/<br/>━━━━━━━━━━<br/>pkg_root · Severity<br/>SkillResolver (Protocol)"]
        HOOK_REG["● hook_registry.py<br/>━━━━━━━━━━<br/>HOOK_REGISTRY<br/>RETIRED_SCRIPT_BASENAMES<br/>compute_registry_hash()"]
        REG_SHA["● registry.sha256<br/>━━━━━━━━━━<br/>expected hash artifact<br/>(sync integrity sentinel)"]
    end

    %% ─── HOOKS LAYER ────────────────────────────────────────────────────── %%
    subgraph HOOKS ["HOOKS — Standalone Subprocess Scripts (stdlib-only)"]
        direction LR
        GUARD["★ grep_pattern_lint_guard.py<br/>━━━━━━━━━━<br/>PreToolUse hook · deny BRE \\|<br/>corrects to ERE | in reason<br/><i>stdlib only — zero autoskillit imports</i>"]
        HOOKS_INIT["hooks/__init__.py<br/>━━━━━━━━━━<br/>re-exports HOOK_REGISTRY<br/>HookDef · generate_hooks_json()"]
    end

    %% ─── L2 RECIPE LAYER ─────────────────────────────────────────────────── %%
    subgraph RECIPE ["L2 — RECIPE (Static Analysis)"]
        direction TB
        RULES_SC["● rules_skill_content.py<br/>━━━━━━━━━━<br/>+ grep-bre-alternation-in-skill (NEW)<br/>no-autoskillit-import · hardcoded-origin<br/>undefined-bash-placeholder"]
        RECIPE_ANALYSIS["recipe/_analysis.py<br/>━━━━━━━━━━<br/>ValidationContext"]
        RECIPE_PARSER["recipe/_skill_placeholder_parser.py<br/>━━━━━━━━━━<br/>extract_bash_blocks()"]
        RECIPE_REG["recipe/registry.py<br/>━━━━━━━━━━<br/>@semantic_rule · RuleFinding<br/>run_semantic_rules()"]
        RECIPE_CONTRACTS["recipe/contracts.py<br/>━━━━━━━━━━<br/>load_bundled_manifest()<br/>resolve_skill_name()"]
    end

    %% ─── TESTS ──────────────────────────────────────────────────────────── %%
    subgraph TESTS ["TESTS"]
        direction TB
        T_INFRA["★ tests/infra/<br/>test_grep_pattern_lint_guard.py<br/>━━━━━━━━━━<br/>12 unit tests for hook guard<br/>patches stdin/stdout"]
        T_SKILLS["★ tests/skills/<br/>test_skill_tool_syntax_contracts.py<br/>━━━━━━━━━━<br/>corpus scan: all SKILL.md bash blocks<br/>parametrized over skills_extended/"]
        T_RECIPE["● tests/recipe/<br/>test_rules_skill_content.py<br/>━━━━━━━━━━<br/>+ test_grep_bre_alternation_is_flagged<br/>+ test_git_grep_bre_is_excluded"]
        T_ARCH["● tests/arch/<br/>test_subpackage_isolation.py<br/>━━━━━━━━━━<br/>hooks/ file-count exemption: 23→24<br/>mentions grep_pattern_lint_guard.py"]
    end

    %% ─── EXTERNAL ───────────────────────────────────────────────────────── %%
    subgraph EXT ["EXTERNAL — stdlib"]
        direction LR
        STDLIB["json · sys · re<br/>hashlib · pathlib · dataclasses<br/>collections.abc · typing"]
    end

    %% ─── VALID DEPENDENCIES ─────────────────────────────────────────────── %%
    HOOK_REG -->|"imports pkg_root"| CORE
    HOOKS_INIT -->|"re-exports"| HOOK_REG
    RULES_SC -->|"imports Severity"| CORE
    RULES_SC -->|"imports ValidationContext"| RECIPE_ANALYSIS
    RULES_SC -->|"imports extract_bash_blocks"| RECIPE_PARSER
    RULES_SC -->|"imports @semantic_rule · RuleFinding"| RECIPE_REG
    RULES_SC -->|"imports load_bundled_manifest"| RECIPE_CONTRACTS

    HOOK_REG -.->|"hash written to"| REG_SHA

    %% Hook guard is registered by hook_registry (name reference, not import) %%
    HOOK_REG -.->|"registers by name"| GUARD

    %% Test dependencies %%
    T_INFRA -->|"imports main()"| GUARD
    T_SKILLS -->|"imports extract_bash_blocks"| RECIPE_PARSER
    T_RECIPE -->|"imports rules_skill_content"| RULES_SC
    T_RECIPE -->|"imports load_recipe · run_semantic_rules"| RECIPE_REG
    T_ARCH -.->|"exemption count bump"| HOOKS_INIT

    %% stdlib %%
    GUARD -->|"stdlib only"| STDLIB
    HOOK_REG -->|"stdlib"| STDLIB
    RULES_SC -->|"re · pathlib"| STDLIB

    %% CLASS ASSIGNMENTS %%
    class CORE stateNode;
    class HOOK_REG handler;
    class REG_SHA output;
    class GUARD newComponent;
    class HOOKS_INIT handler;
    class RULES_SC newComponent;
    class RECIPE_ANALYSIS,RECIPE_PARSER,RECIPE_CONTRACTS phase;
    class RECIPE_REG stateNode;
    class T_INFRA,T_SKILLS newComponent;
    class T_RECIPE,T_ARCH detector;
    class STDLIB integration;
```

### Scenarios Diagram

```mermaid
%%{init: {'flowchart': {'nodeSpacing': 40, 'rankSpacing': 50, 'curve': 'basis'}}}%%
flowchart LR
    %% CLASS DEFINITIONS %%
    classDef cli fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;
    classDef stateNode fill:#004d40,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef handler fill:#e65100,stroke:#ffb74d,stroke-width:2px,color:#fff;
    classDef phase fill:#6a1b9a,stroke:#ba68c8,stroke-width:2px,color:#fff;
    classDef newComponent fill:#2e7d32,stroke:#81c784,stroke-width:2px,color:#fff;
    classDef output fill:#00695c,stroke:#4db6ac,stroke-width:2px,color:#fff;
    classDef detector fill:#b71c1c,stroke:#ef5350,stroke-width:2px,color:#fff;
    classDef gap fill:#ff6f00,stroke:#ffa726,stroke-width:2px,color:#000;
    classDef terminal fill:#1a237e,stroke:#7986cb,stroke-width:2px,color:#fff;

    subgraph S1 ["SCENARIO 1: BRE Pattern Interception (Primary Happy Path)"]
        direction LR
        S1_MODEL["Claude Model<br/>━━━━━━━━━━<br/>emits Grep tool call<br/>pattern: 'foo\\|bar'"]
        S1_HOOK["★ grep_pattern_lint_guard.py<br/>━━━━━━━━━━<br/>PreToolUse hook<br/>detects \\| sentinel"]
        S1_DENY["deny + corrected pattern<br/>━━━━━━━━━━<br/>permissionDecision: deny<br/>corrected: 'foo|bar'"]
        S1_RETRY["Model retries<br/>━━━━━━━━━━<br/>Grep pattern: 'foo|bar'<br/>ERE — ripgrep succeeds"]
    end

    subgraph S2 ["SCENARIO 2: Recipe Rule Flags Skill BRE Usage"]
        direction LR
        S2_RECIPE["Recipe Validator<br/>━━━━━━━━━━<br/>validate_recipe() runs<br/>semantic rules"]
        S2_RULE["● rules_skill_content.py<br/>━━━━━━━━━━<br/>grep-bre-alternation-in-skill<br/>rule checks bash blocks"]
        S2_PARSER["_skill_placeholder_parser<br/>━━━━━━━━━━<br/>extract_bash_blocks()<br/>reads SKILL.md content"]
        S2_FINDING["RuleFinding: ERROR<br/>━━━━━━━━━━<br/>severity=ERROR<br/>violations list emitted"]
    end

    subgraph S3 ["SCENARIO 3: Hook Registry Wiring"]
        direction LR
        S3_REG["● hook_registry.py<br/>━━━━━━━━━━<br/>HOOK_REGISTRY adds<br/>HookDef(matcher='Grep')"]
        S3_JSON["generate_hooks_json()<br/>━━━━━━━━━━<br/>writes hooks.json with<br/>grep_pattern_lint_guard"]
        S3_DRIFT["_count_hook_registry_drift()<br/>━━━━━━━━━━<br/>detects missing hooks<br/>vs deployed settings.json"]
        S3_SHA["● registry.sha256<br/>━━━━━━━━━━<br/>HOOK_REGISTRY_HASH<br/>integrity sentinel"]
    end

    subgraph S4 ["SCENARIO 4: Skill Syntax Contract Testing"]
        direction LR
        S4_TEST["★ test_skill_tool_syntax_contracts.py<br/>━━━━━━━━━━<br/>@parametrize all skill dirs<br/>in skills/ + skills_extended/"]
        S4_EXTRACT["extract_bash_blocks()<br/>━━━━━━━━━━<br/>reads each SKILL.md<br/>extracts bash fences"]
        S4_SKILLS["● audit-friction/SKILL.md<br/>● dry-walkthrough/SKILL.md<br/>● merge-pr/SKILL.md<br/>━━━━━━━━━━<br/>grep BRE patterns removed"]
        S4_PASS["assert violations == []<br/>━━━━━━━━━━<br/>no grep '\\|' in bash blocks<br/>test passes"]
    end

    subgraph S5 ["SCENARIO 5: Guard Allow Pass-Through"]
        direction LR
        S5_ERE["Grep call with 'foo|bar'<br/>━━━━━━━━━━<br/>ERE alternation — valid<br/>no \\| sentinel present"]
        S5_NONGREP["Non-Grep tool call<br/>━━━━━━━━━━<br/>Bash, Read, Edit, etc.<br/>tool_name != 'Grep'"]
        S5_MALFORMED["Malformed JSON stdin<br/>━━━━━━━━━━<br/>JSONDecodeError path<br/>fail-open design"]
        S5_ALLOW["★ grep_pattern_lint_guard.py<br/>━━━━━━━━━━<br/>sys.exit(0) — no output<br/>Claude Code allows call"]
    end

    %% SCENARIO 1 FLOW %%
    S1_MODEL -->|"reads pattern"| S1_HOOK
    S1_HOOK -->|"writes deny + reason"| S1_DENY
    S1_DENY -->|"passes corrected pattern"| S1_RETRY

    %% SCENARIO 2 FLOW %%
    S2_RECIPE -->|"loads skill via run_skill step"| S2_RULE
    S2_RULE -->|"reads SKILL.md content"| S2_PARSER
    S2_PARSER -->|"returns bash blocks"| S2_RULE
    S2_RULE -->|"writes finding"| S2_FINDING

    %% SCENARIO 3 FLOW %%
    S3_REG -->|"passes HookDef list"| S3_JSON
    S3_JSON -->|"writes hooks.json"| S3_SHA
    S3_REG -->|"reads canonical basenames"| S3_DRIFT
    S3_SHA -->|"validates integrity"| S3_DRIFT

    %% SCENARIO 4 FLOW %%
    S4_TEST -->|"reads each SKILL.md"| S4_EXTRACT
    S4_EXTRACT -->|"returns bash blocks"| S4_TEST
    S4_TEST -->|"checks against"| S4_SKILLS
    S4_SKILLS -->|"no violations"| S4_PASS

    %% SCENARIO 5 FLOW %%
    S5_ERE -->|"reads pattern — no \\| found"| S5_ALLOW
    S5_NONGREP -->|"reads tool_name — exits early"| S5_ALLOW
    S5_MALFORMED -->|"JSONDecodeError — exits early"| S5_ALLOW

    %% CLASS ASSIGNMENTS %%
    class S1_MODEL,S5_ERE,S5_NONGREP,S5_MALFORMED cli;
    class S1_HOOK,S5_ALLOW newComponent;
    class S1_DENY,S2_FINDING output;
    class S1_RETRY,S3_SHA stateNode;
    class S2_RECIPE,S4_TEST newComponent;
    class S2_RULE,S3_REG handler;
    class S2_PARSER,S4_EXTRACT phase;
    class S3_JSON,S3_DRIFT phase;
    class S4_SKILLS,S4_PASS output;
```

Closes #1184

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/remediation-20260424-134738-177468/.autoskillit/temp/rectify/rectify_grep-bre-syntax-immunity_2026-04-24_135844_part_a.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | uncached | output | cache_read | cache_write | count | time |
|------|----------|--------|------------|-------------|-------|------|
| investigate | 102 | 8.9k | 392.3k | 34.4k | 1 | 5m 19s |
| rectify | 113 | 24.2k | 589.7k | 68.6k | 1 | 10m 26s |
| dry_walkthrough | 312 | 26.0k | 2.1M | 140.7k | 2 | 6m 54s |
| implement | 472 | 21.2k | 2.9M | 105.7k | 2 | 7m 2s |
| assess | 190 | 7.3k | 982.8k | 46.8k | 1 | 3m 46s |
| prepare_pr | 60 | 6.1k | 205.8k | 28.9k | 1 | 1m 54s |
| run_arch_lenses | 199 | 18.1k | 716.3k | 117.3k | 3 | 7m 9s |
| compose_pr | 67 | 9.6k | 260.3k | 35.9k | 1 | 2m 17s |
| **Total** | 1.5k | 121.4k | 8.2M | 578.5k | | 44m 50s |

### CODE REVIEW

File: tests/infra/test_grep_pattern_lint_guard.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+
+def _decision(result: dict | None) -> str:
+    if result is None:
+        return "allow"
+    return result.get("hookSpecificOutput", {}).get("permissionDecision", "allow")
+
+
+def _reason(result: dict | None) -> str:
+    return result.get("hookSpecificOutput", {}).get("permissionDecisionReason", "")
```
Comment (Trecek) [inline tests/infra/test_grep_pattern_lint_guard.py:32]: [warning] bugs: `_reason()` does not guard against `result` being `None` — calling `result.get(...)` raises `AttributeError` when `result` is `None` (the allow path). Fix: add `if result is None: return ""` at the top of the function, matching the pattern used by `_decision()`.

File: src/autoskillit/hooks/grep_pattern_lint_guard.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    try:
+        data = json.loads(sys.stdin.read())
+    except (json.JSONDecodeError, ValueError, OSError):
+        sys.exit(0)  # fail-open on malformed input
+
+    tool_name = data.get("tool_name", "")
+    if tool_name != "Grep":
+        sys.exit(0)  # guard is Grep-specific
+
+    pattern = data.get("tool_input", {}).get("pattern", None)
```
Comment (Trecek) [inline src/autoskillit/hooks/grep_pattern_lint_guard.py]: [warning] bugs: If the hook receives JSON where `tool_input` is an explicit `null` (rather than omitted), `data.get("tool_input", {})` returns `None` instead of `{}`, and the chained `.get("pattern", None)` raises `AttributeError`. This is not caught by the `except (json.JSONDecodeError, ValueError, OSError)` block, so the script exits non-zero — blocking the Grep call rather than failing open. Fix: change to `(data.get("tool_input") or {}).get("pattern", None)` or add `AttributeError` to the caught exception types.

File: src/autoskillit/skills_extended/merge-pr/SKILL.md [hunk 3]
```diff
@@ -139,7 +139,7 @@ git log --diff-filter=D --oneline --follow -- {file_path} origin/${BASE_BRANCH}
 
 # For each regressed symbol: find the commit that removed it
 git log --diff-filter=M --oneline -p -- {file_path} origin/${BASE_BRANCH} \
-  | grep -B20 "^-def {symbol_name}\|^-class {symbol_name}" \
+  | rg -B 20 "^-def {symbol_name}|^-class {symbol_name}" \
   | grep "^[0-9a-f]\{7,\}" \
```
Comment (Trecek) [inline src/autoskillit/skills_extended/merge-pr/SKILL.md]: [warning] cohesion: After migrating from `grep -B20 "...\\|..."` to `rg -B 20`, the next pipeline stage `| grep "^[0-9a-f]\\{7,\\}"` still uses POSIX BRE brace syntax and `grep`. This is inconsistent with the migration goal. Convert to `rg "^[0-9a-f]{7,}"` for consistency.

File: src/autoskillit/recipe/rules_skill_content.py [hunk 4]
```diff
@@ -254,6 +254,85 @@ def _check_no_autoskillit_import(ctx: ValidationContext) -> list[RuleFinding]:
...
+    description=(
+        "A SKILL.md bash block uses grep with BRE \\| alternation. "
+        "The Grep tool wraps ripgrep (ERE) where | (bare) is alternation. "
+        "Models copying \\| from skill bash blocks into Grep tool calls get 0 results silently. "
+        "Fix: replace grep 'foo\\|bar' with rg 'foo|bar'. "
+        "Exception: --grep= arguments in git log/show commands are legitimate BRE."
+    ),
+)
+def _check_no_grep_bre_alternation(ctx: ValidationContext) -> list[RuleFinding]:
+    """Flag bash blocks in SKILL.md that use grep \\| BRE alternation.
```
Comment (Trecek) [inline src/autoskillit/recipe/rules_skill_content.py]: [warning] slop: The docstring for `_check_no_grep_bre_alternation` fully restates the content of the `description=` argument in the `@semantic_rule` decorator immediately above it. Remove the docstring or reduce it to a single non-redundant sentence — the decorator already surfaces this text as the rule description.

File: tests/skills/test_skill_tool_syntax_contracts.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    result = []
+    for skill_dir in _SKILL_DIRS:
+        if not skill_dir.exists():
+            continue
+        result.extend(p for p in sorted(skill_dir.iterdir()) if p.is_dir())
+    return result
+
+
+# Regex that matches a standalone grep command line (not --grep= option) containing \|
+_GREP_BRE_ALTERNATION = re.compile(
```
Comment (Trecek) [inline tests/skills/test_skill_tool_syntax_contracts.py]: [warning] cohesion: `_GREP_BRE_ALTERNATION` and `_GIT_GREP_BRE` duplicate `_GREP_BRE_ALTERNATION_RE` and `_GIT_GREP_BRE_RE` from `src/autoskillit/recipe/rules_skill_content.py`. If the patterns diverge, the hook guard, semantic rule, and this test enforce different grammars silently. Import the compiled patterns from the rules module to prevent drift. Also: naming inconsistency — test uses no `_RE` suffix while the source does.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: warning-only findings detected. See inline comments — no blocking changes required.

File: tests/infra/test_grep_pattern_lint_guard.py (hunk 1 above)
Comment (Trecek) [reply tests/infra/test_grep_pattern_lint_guard.py:32]: Agreed — fixed in 66ecc4a1. Added `if result is None: return ""` guard to `_reason()` matching the pattern already used in `_decision()`.

File: src/autoskillit/hooks/grep_pattern_lint_guard.py (hunk 2 above)
Comment (Trecek) [reply src/autoskillit/hooks/grep_pattern_lint_guard.py]: Agreed — fixed in ba316a3f. Changed `data.get("tool_input", {})` to `(data.get("tool_input") or {})` so an explicit JSON `null` value is treated as absent rather than causing `AttributeError` on the chained `.get()`.

File: src/autoskillit/skills_extended/merge-pr/SKILL.md (hunk 3 above)
Comment (Trecek) [reply src/autoskillit/skills_extended/merge-pr/SKILL.md]: Agreed — fixed in 1fe02ac0. Replaced `grep "^[0-9a-f]\\{7,\\}"` with `rg "^[0-9a-f]{7,}"` so the entire pipeline stage is consistently ripgrep.

File: src/autoskillit/recipe/rules_skill_content.py (hunk 4 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules_skill_content.py]: Agreed — fixed in f9cc6f85. Removed the docstring body that fully restated the `description=` argument in the `@semantic_rule` decorator above it.

File: tests/skills/test_skill_tool_syntax_contracts.py (hunk 5 above)
Comment (Trecek) [reply tests/skills/test_skill_tool_syntax_contracts.py]: Agreed — fixed in 021ed1e7. Replaced the locally-defined `_GREP_BRE_ALTERNATION` and `_GIT_GREP_BRE` regex copies with direct imports of `_GREP_BRE_ALTERNATION_RE` and `_GIT_GREP_BRE_RE` from `rules_skill_content`, eliminating the drift risk.
